ConcurrentAPI.adoc

```diff
@@ -0,0 +1,255 @@
+   boolean defineLock(String name, LockConfiguration configuration);
+   InfinispanLock get(String name);
+   LockConfiguration getConfiguration(String name);
+   boolean isDefined(String name);
+   CompletableFuture<Boolean> remove(String name);
+   CompletableFuture<Void>	unlock(String name);
+}
+
+
+public class LockConfiguration {
+   private final RentrancyLevel lentrancyLevel; // default None
+   private final boolean silentFailover; // default true
+   private final long maxLeaseTime; // after a time the lock is released in milles
+   private final long renewalLeaseTime; // time after a client can aquire the lock
+}
+
+public enum RentrancyLevel {
+   None, Node, Thread
+}
+
+```
+###### InfinispanLock defineLock(String name, LockConfiguration configuration)
+
+Defines a lock with the specific name and LockConfiguration.
+It does not overwrite existing configurations.
+
+Returns true if successfully defined or false if the lock is already defined or any other failure.
+If silentFailover is false, then InfinispanLockException will be raised.
+
+###### InfinispanLock get(String name)
+
+Get's a InfinipanLock by it's name and throws InfinispanLockException if the lock is not not defined.
+User must call defineLock before this method.
+
+###### Optional<LockConfiguration> getConfiguration(String name);
+
+Get's the configuration for a Lock. If the Lock does not exist, Optional.empty() will be returned.
+
+###### boolean isDefined(String name)
+True if the lock exists, false if it doesn't
+
+###### CompletableFuture<Boolean> remove(String name)
+Removes a Lock from the system. Returns true when it was removed, false when the lock does not exist.
+If any other Runtime problems appear, InfinispanLockException will be raised withe the reason.
+As Locks are not removed automatically, so this has to be done programatically when the Lock is
+no longer needed. Otherwise, OutOfMemoryException could happen.
+
+Remove should be executed when the lock is locked, because running that without exclusive access should
+result in an exception.
+Internally, the implementation should contain generation number so that attempts to acquire a lock
+of a removed generation will result it exceptions in the other callers, too.
+
+###### CompletableFuture<Void> unlock(String name)
+Unlocks a lock in the system. This might be useful to unlock a lock from a management point of view if
+needed to.
+
+=== InfinispanLock
+
+When a instance holding a Lock dies, this lock is released and available for the other instances.
+
+```java
+public interface InfinispanLock {
+
+   CompletableFuture<Void>	lock();
+
+   CompletableFuture<Boolean>	tryLock();
+
+   CompletableFuture<Boolean>	tryLock(long time, TimeUnit unit);
+
+   CompletableFuture<Void>	unlock();
+}
+
+```
+
+###### CompletableFuture<Void> lock();
+
+CompletableFuture is completed successfully when the lock is acquired.
+In client/server mode, when a lock is aquired by a client, it will be automatically released
```

Comment:
In other words, I would make the defaults `-1` and support only those in initial embedded implementation, but later the remote client should not support non-positive (because that would lead to deadlocks).